tox_conda/plugin.py

```diff
@@ -1,332 +1,419 @@
+        finally:
+            tear_down()
+
+        install_command = CondaEnvRunner._generate_install_command(
+            conda_exe, python, conda_cache_conf
+        )
+        if install_command:
+            try:
+                install_command_args = shlex.split(install_command)
+                subprocess.run(install_command_args, check=True)
+            except subprocess.CalledProcessError as e:
+                raise Fail(f"Failed to install dependencies in conda environment. Error: {e}")
+
+    @staticmethod
+    def _generate_env_create_command(conda_exe, python, conda_cache_conf):
+        env_path = Path(conda_cache_conf["env_path"]).resolve()
         # conda env create does not have a --channel argument nor does it take
         # dependencies specifications (e.g., python=3.8). These must all be specified
         # in the conda-env.yml file
         yaml = YAML()
         env_file = yaml.load(env_path)
         env_file["dependencies"].append(python)
 
-        tmp_env = tempfile.NamedTemporaryFile(
+        tmp_env_file = tempfile.NamedTemporaryFile(
             dir=env_path.parent,
             prefix="tox_conda_tmp",
             suffix=".yaml",
             delete=False,
         )
-        yaml.dump(env_file, tmp_env)
-
-        args = [
-            venv.envconfig.conda_exe,
-            "env",
-            "create",
-            "-p",
-            envdir,
-            "--file",
-            tmp_env.name,
-        ]
-        tmp_env.close()
-        _run_conda_process(args, venv, action, basepath)
-        Path(tmp_env.name).unlink()
-
-    else:
-        args = [venv.envconfig.conda_exe, "create", "--yes", "-p", envdir]
-        for channel in venv.envconfig.conda_channels:
-            args += ["--channel", channel]
-
-        # Add end-user conda create args
-        args += venv.envconfig.conda_create_args
-
-        args += [python]
-
-        _run_conda_process(args, venv, action, basepath)
-
-    venv.envconfig.conda_python = python
-
-    # let the venv know about the target interpreter just installed in our conda env, otherwise
-    # we'll have a mismatch later because tox expects the interpreter to be existing outside of
-    # the env
+        yaml.dump(env_file, tmp_env_file)
+        tmp_env_file.close()
+
+        cmd = f"'{conda_exe}' env create --file '{tmp_env_file.name}' --quiet --force"
+        tear_down = lambda: Path(tmp_env_file.name).unlink()
+
+        return cmd, tear_down
+
+    @staticmethod
+    def _generate_create_command(conda_exe, python, conda_cache_conf):
+        cmd = f"'{conda_exe}' create {conda_cache_conf['env_spec']} '{conda_cache_conf['env']}' {python} --yes --quiet"
+        for arg in conda_cache_conf.get("create_args", []):
+            cmd += f" '{arg}'"
+
+        tear_down = lambda: None
+        return cmd, tear_down
+
+    @staticmethod
+    def _generate_install_command(conda_exe, python, conda_cache_conf):
```

Comment:
Can we add type annotations universally, please.

Reply:
I have added type info to all function defs.